### PR TITLE
[Model] Pipeline Parallel Support for DeepSeek v2

### PR DESCRIPTION
**Pull request by tjohnson31415**
Adds pipeline parallel support for DeepSeek v2.

Tested with https://huggingface.co/deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct using `--tensor-parallel-size 1 --pipeline-parallel-size 2`

### CODE REVIEW

**Comment by youkaichao** [comment]
can you test the correctness locally, using https://github.com/vllm-project/vllm/blob/main/tests/distributed/test_pipeline_parallel.py ?

**Comment by tjohnson31415** [comment]
> can you test the correctness locally, using https://github.com/vllm-project/vllm/blob/main/tests/distributed/test_pipeline_parallel.py ?

Sure. I edited the file to set the model to `"deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct"`and added `--trust-remote-code` to the launch args and was able to see the `test_pipeline_parallel` tests passing:
```
$ pytest -s tests/distributed/test_pipeline_parallel.py
```
```
...
.INFO:     Shutting down


=============================== warnings summary ===============================
my-vllm/lib64/python3.11/site-packages/transformers/utils/hub.py:127
  /workspace/my-vllm/lib64/python3.11/site-packages/transformers/utils/hub.py:127: FutureWarning: Using `TRANSFORMERS_CACHE` is deprecated and will be removed in v5 of Transformers. Use `HF_HOME` instead.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================== 5 passed, 1 warning in 393.58s (0:06:33) ===================
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [75971]
INFO 07-18 15:40:48 multiproc_worker_utils.py:136] Terminating local vLLM worker processes
(VllmWorkerProcess pid=76043) INFO 07-18 15:40:48 multiproc_worker_utils.py:237] Worker exiting
INFO 07-18 15:40:48 async_llm_engine.py:51] Engine is gracefully shutting down.
[rank0]:[W CudaIPCTypes.cpp:16] Producer process has been terminated before all shared CUDA tensors released. See Note [Sharing CUDA tensors]
/usr/lib64/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
  ```

**Comment by tjohnson31415** [comment]
Rebased to resolve conflict from `main`. Reran the tests and they pass still.

File: vllm/model_executor/models/deepseek_v2.py [hunk 1]
```diff
@@ -394,33 +398,56 @@ def __init__(
...
+        else:
+            self.embed_tokens = PPMissingLayer()
+
+        self.start_layer, self.end_layer, self.layers = make_layers(
+            config.num_hidden_layers,
+            # layer_idx is still an argument
+            functools.partial(DeepseekV2DecoderLayer,
+                              config,
+                              cache_config=cache_config,
+                              quant_config=quant_config),
```
Comment (youkaichao) [inline vllm/model_executor/models/deepseek_v2.py:415]: this lambda function will have `prefix=` shortly after https://github.com/vllm-project/vllm/pull/6515 .

File: vllm/model_executor/models/deepseek_v2.py [hunk 2]
```diff
@@ -389,38 +413,62 @@ def __init__(
...
-            config.hidden_size,
-        )
-        self.layers = nn.ModuleList([
-            DeepseekV2DecoderLayer(config,
-                                   layer_idx,
-                                   cache_config=cache_config,
-                                   quant_config=quant_config)
-            for layer_idx in range(config.num_hidden_layers)
-        ])
-        self.norm = RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
```
Comment (youkaichao) [inline vllm/model_executor/models/deepseek_v2.py:408]: I suggest change this class to accept `prefix=` , and parse `layer_index` inside.

File: vllm/model_executor/models/utils.py [hunk 3]
```diff
@@ -140,11 +141,18 @@ def make_layers(
...
+    ])
     modules = torch.nn.ModuleList(
-        [PPMissingLayer() for _ in range(start_layer)] + [
-            maybe_offload_to_cpu(layer_fn(prefix=f"{prefix}.{idx}"))
-            for idx in range(start_layer, end_layer)
-        ] + [PPMissingLayer() for _ in range(end_layer, num_hidden_layers)])
+        [PPMissingLayer() for _ in range(start_layer)] +
+        [maybe_offload_to_cpu(layer_module)
+         for layer_module in layer_modules] +
+        [PPMissingLayer() for _ in range(end_layer, num_hidden_layers)])
```
Comment (youkaichao) [inline vllm/model_executor/models/utils.py:155]: should be able to avoid change in this file.

**Review by youkaichao — APPROVED** [review]
thanks for addressing my comments! please test the correctness locally.

**Comment by youkaichao** [comment]
Thanks, that might be caused by the flakiness of pp tests. I'll merge as this PR looks good to me now.

Thanks for your contribution!

**Pull request merged**